### DISCUSSION TITLE
CDA JavaScript/TypeScript Client-Side SDK

### DISCUSSION THREAD

**Discussion by krowvin** [USACE/cwms-data-api#351] [Ideas]
# CWMS Data API Software Development Kit

*Posting here to open discussion on the creation of a browser based software development kit for CDA*

Ideally the software would be modeled after the existing DBAPI python api. This would give users of CWMS that already use DBAPI an advantage. 

Some icebreakers:
1. Open source -  Both public and internal could use this
2. Provides users with a way to focus on their app and not on what `fetch()` and/or `XHR` are, what api version we are on, etc 
3. Could manage the handling of tokens once auth is completed
4. Has all the endpoints
5. NPM package for ease of use with package managers/projects
6. Would update as changes are made
7. Wraps the DBAPI as much as possible

## Name Suggestions
* CDA-SDK
* CDAjs
* ??

Ideally it would be written with typescript, and could be used by both javascript and typescript projects. 

Some examples would be: 

### Getting a catalog of locations
```
const location = new Location(“SWT”)
console.log(await location.catalog())
```

### Fetching a specific timeseries ID for the last 24 hours
```
const timeseries = new Timeseries(“SPK”)
const ts = await timeseries.get(“tsid”, begin, end, etc).then(d=>do something)
for (const t of ts.values) {
   console.log(t.name)
}
```

Other Notes: In theory this could work just fine with NodeJS for server side calls as well.

**Comment by adamkorynta** [comment]
DBAPI only handles time series and ratings so you might find using that for a template against all endpoints will shoehorn you into certain patterns.

**Comment by MikeNeilson** [comment]
I like it. Though I think a factory style would work better.

Something like 
```
const api = cwmsdata.for("https://.....");
const ts = api.timeseries("OFFICE");
```
could even have
```
const api cwmsdata.for(...).withAuth(<creds>);
```
That will make it easier to alter exactly *which* CDA instance is being targeted.

I'll also 2nd what Adam said. I don't think a similar API is a bad idea, but it should be a layer in front of an API better suited to javascript.

**Reply by krowvin** [reply]
Oh yes that's a good point, @adamkorynta @MikeNeilson.

I was thinking users are familiar with the "get" and "put" mentality that is brought on with DBAPI. 

I had a really basic tool fleshed out for my own use. Got tired of typing fetch. Not saying I would do it this way for this effort. Just that I had *[something](https://github.com/USACE-WATER-CONTROL-WEB/Hydrologic-Query-Tool/blob/main/public/js/Radar.js)*. 

RMA was working on a Jython SDK of sorts for CMA I thought. This would not overlap that effort would it?

**Reply by MikeNeilson** [reply]
I'm assuming you meant to type CDA there. and yes, and that specifically IS to mimic the DBAPI as a drop in replacement. 

If a district is going so far as to use javascript/node/etc. Time to modernize design constructs and not bring that much cruft.

**Reply by krowvin** [reply]
> I'm assuming you meant to type CDA there  

I did, CMA is on my mind these days. 

> If a district is going so far as to use javascript/node/etc   

If more districts started to use CDA in a JavaScript environment (See last weeks web training). This could help with that. But doing it up to standards is why I ask here. Many ways to go about this. Am I even the right person or should we contract this task out? Is there a need beyond my desire to only type `fetch` _once_ and use promises elsewhere, for example.

**Reply by adamkorynta** [reply]
The great thing about DBAPI is that engineers who do not want to delve deep into enterprise application patterns and lifecycles have a simpler mechanism to work against. Additionally, it is a library specifically marked as for use by scripters so changes need to be broadcasted differently. Having a JS library that hides what REST/fetch/Cookies/etc certainly seems beneficial. It also lets bigger server-side changes happen as the public JS API would likely remain the same, even if the actual REST API's changed underneath.

**Reply by krowvin** [reply]
> It also lets bigger server-side changes happen as the public JS API would likely remain the same, even if the actual REST API's changed underneath.

100%  

We had a web meetup last week where we made a webpage using react for WM sites.  

The #1 complaint I heard from folks was how terse fetch/async/getting data was. It's certainly doable. But something you can have intellisense on _with_ code examples baked in and documentation would obfuscate away the mundane. Middle users, arguably, do not care about the innerworkings as you eluded to. They just want it. 

Pair it with npm and an update is as easy as `npm i cwmsjs` *name pending*

**Reply by MikeNeilson** [reply]
the downside is that they don't work within any patterns and figuring out what's wrong is a giant mess.

I say go for it for now. At the least a prototype that includes the current catalogs, timeseries, and location data. Maybe location levels if it's going quickly. That's enough to let us review the API itself and comment but not enough that it's annoying to change afterwards or even take too long to prototype.

**Reply by adamkorynta** [reply]
Luckily the #1 issue with DBAPI that we keep hearing about is Connection starvation, and that wouldn't be an issue here.

**Reply by MikeNeilson** [reply]
Famous last words. 😄

**Reply by krowvin** [reply]
Browser limits async requests to 6 as @DanielTOsborne and I have had fun dealing with.

But as more use it scaling will be fun! ☁️

**Reply by MikeNeilson** [reply]
huh, I guess that's not too surprising. Though I would've though async requests can go forever... they're async. Or is that specifically if you use "await."

**Reply by krowvin** [reply]
I think it's actually a limit imposed _per_ browser. Looking at the spec itself HTTP 1.1 may have less, but [2.0 appears to have no limit](https://httpwg.org/specs/rfc7540.html#SETTINGS_MAX_CONCURRENT_STREAMS). 

As for `fetch` and `await`, in javascript, i have only seen them used to access a [Promise](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise) which is what `fetch` returns (if you don't chain). But these let you fire off as many ajax requests as you like in the browser. But only 6 will complete at a time (for Chrome). 

All this is moot if you are talking about any other async and/or language. Just talking about browser based apps (which is what the proposed SDK here would be for primarily)

**Comment by MikeNeilson** [comment]
Also that seems like the kind of thing to put in this repo. since it needs to stay up-to-date anyways. There are several OpenAPI to "client impl" libraries available so it'd mostly be the hec-cwms-data-access esque glue to keep up-to-date.

**Reply by krowvin** [reply]
If there's a more automated way to transpile for this, i'm all for it! I realize the endpoints are still fluctuating (BETA).

**Reply by MikeNeilson** [reply]
It's not a transpile. just generation. https://openapi-generator.tech/docs/generators/javascript/

Given our current design it would need to run after the integration tests or, at least it needs to run the "run" task to have an instance.

**Reply by adamkorynta** [reply]
I see usefulness if keeping an autogenerated library within this repo, but I think you'd want to keep the public interface/factories/etc separate so that you aren't mixing server-side PR's with client side updates and vice versa.

**Reply by MikeNeilson** [reply]
good point. and that is how the java version currently works.

**Comment by adamkorynta** [comment]
Using a Fluent API style for this would really make nice examples for users. That's something the DBAPI is lacking.

**Reply by MikeNeilson** [reply]
Agreed.

**Comment by rma-rripken** [comment]
The Jython API for CDA that you were referring to is RADARAPI.  Its just the time series and rating methods from DBAPI but backed by CDA instead of JDBC.   I think your idea is different.

**Reply by krowvin** [reply]
I'm still not %100 what all DBAPI does. I intended on looking into it and copying method names. I was just under the impression it obfuscated the port/connection/etc away from users. 

But realizing from the other thread that maybe *this* should be its own thing. 

I also wanted to make sure someone wasn't **already** doing this. Because I'd just contribute to their project. 

I look forward to doing was RADARAPI has. It's geared towards jython/python iirc?

Whereas I was intending this as something for browser/client apps.

**Comment by rma-rripken** [comment]
The RADARAPI write side of things is still being developed - I've been held back b/c I've had difficulties setting up the docker integration tests.   Mainly its just calling the java cda client code from jython and translating objects

**Comment by krowvin** [comment]
Just to update folks on where this is. Neilson created a repo here:
https://github.com/HydrologicEngineeringCenter/cwms-data-api-client-javascript/tree/openapi-generator

I was able to generate TypeScript and JavaScript client libraries using a trimmed version of the Swagger Docs API. I need to do more testing with the library and have had issues importing it (something to do with modules / ES6). But it appears to be a solid route and would enable us to integrate it in the build automation for future updates of CDA.

**Comment by krowvin** [comment]
@rma-rripken @MikeNeilson 

When implementing the swagger doc into the openapi generator one issue I ran into was that we have the api path (/cwms-data) added to _all_ the endpoints. 

This makes it very difficult to override this in the generated library for the end user.  
Looking through the openapi generated source it appears the `path` is not directly exposed to the configuration only the `basePath` is. 

I propose we remove the `/cwms-data` basePath from the swagger config and add it into the [host](https://github.com/USACE/cwms-data-api/blob/1957690bd628a8059baa8659946f16d6507ea2f6/cwms-data-api/src/main/webapp/swagger-config.yaml#L10). 

I manually made this change to my `swagger-config.json` and ran the generator to confirm it worked. 

It is also possible this is a limitation of the generator.  
But looking at the generated source [it appears to only have a basePath](https://github.com/HydrologicEngineeringCenter/cwms-data-api-client-javascript/blob/6d3069e173becabb6a6b1671ae0f57dacaf98677/src/src/runtime.ts#L18).  
There is a `path` but it looks like it would take extra steps to set the `RequestOpts` [here](https://github.com/HydrologicEngineeringCenter/cwms-data-api-client-javascript/blob/6d3069e173becabb6a6b1671ae0f57dacaf98677/src/src/runtime.ts#L144).  

Doing this would alter my #588 to change the servers to [this](https://github.com/HydrologicEngineeringCenter/cwms-data-api-client-javascript/blob/6d3069e173becabb6a6b1671ae0f57dacaf98677/cwms-swagger.json#L10)  
or  
`- url: https://cwms-data.usace.army.mil/cwms-data`

Example client source after the change:
```javascript
import { OFFICE, HOURLY_INST } from "../js/constants.js";
import { Configuration, TimeSeriesApi } from "cwmsjs";
import { getDate } from "../js/utils"

const CONF = new Configuration({
  basePath: "https://t7-hostname:tomcatport/district-data",
  headers: { accept: "application/json;version=2" },
});

const TSA = new TimeSeriesApi(CONF);
TSA 
  .getCwmsDataTimeseries({
    name: ts,
    office: OFFICE,
    begin: getDate(-5, `YYYY-MM-DDT${HOURLY_INST.time}:00Z`),
  })
  .then(
    (data) => {
      console.log("API called successfully. Returned data: ", data);
      setTSValues(data?.values || []);
    },
    (error) => {
      console.error(error);
    }
  );
```
  
This also could be a knowledge issue on my part as the auto generated documentation also needs some help. (May end up writing it manually, `typescript` will generate all the docs but `typescript-fetch` will only make a Readme. )

On the docs, one option is to use mustache/templates.  [typescript-fetch generator docs](https://openapi-generator.tech/docs/generators/typescript-fetch)
